### PR TITLE
feat: Startup scripts & Docker images

### LINKED ISSUE #82: Reproducible installation of full system via docker [8]
Simplify the compilation/installation as much as possible, so as a user, i can get the system running on Linux with one script.

Ubuntu 20.04

### PR DESCRIPTION
**Pull request by Gozzim**
## Changes Proposed:

- Merge all Dockerfiles into one
- Proper images for the separate packages and services in Docker
- Properly adjusted docker-compose services
- Added Demo Docker Image
- Added Demo script in root package.json
- Added scripts to start the services and increase usability
- Various fixes

## Issues Addressed:

- Closes #82

## Tests Performed:

- Compiled and started every single docker image
- Started the demo natively
- Tested the StartUp script (./cdtcloud)

## How to Test the Changes:

- Requirements:
  - Running on Host-System
    - NodeJS Version 16.5 or higher
    - yarn
    - Postgresql DB server
  - Running in Docker: 
    - docker
    - docker-compose (Supporting Version 3.7 or higher)
- Start with Docker (Option 1)
  1. Run `./cdtcloud`
  2. Select the `Run docker tools` option
  3. Run the demo
  4. Also test every image by running the services separately
- Start Native on Host-System (Option 2)
  1. Run `./cdtcloud` (needs to be run as Admin on Windows)
  2. Select the `First installation` option
  3. Run the `Run demo` option
- Testing
  - Go to http://localhost:3000 for the theia browser implementation
  - Select View -> Cdtcloud Widget to open the extension
  - Create or open a file (An example project like [this](https://create.arduino.cc/projecthub/EvdS/led-dice-885cf1) can be used for testing purposes)
  - Attach a supported device (for example an Arduino Uno)
  - The widget list should now show a selectable board
  - Hit deploy to compile and upload the code onto the selected device
  - Go to http://localhost:3001 to view the Admin Dashboard

## Known Issues and TODO List:

- [x] Arduino-cli installation via script and startup command
- [x] Database isn't automatically initiated yet on the first run. Currently needs to be executed separately.
- [ ] Running the arduino-cli in docker on windows does not recognize attached devices due to the `privileged` option not being supported.
- [ ] When building theia for the browser, the electron app sometimes does not work anymore. The same problem occurs the other way around. 

## How to Test PRs
1) First fetch the Pull-Request from the main repo, replacing `xx` with the ID of the PR:\
`git fetch git@github.com:eclipsesource/cdtcloud-deploymentserver.git pull/xx/head:pr_xx`
2) Checkout the Pull-Request, again replacing `xx` with the PR-ID:\
`git checkout pr_xx`
3) Perform the necessary tests for the PR

### CODE REVIEW

**Comment by Gozzim** [comment]
@theZasa I would appreciate an extensive test from you specifically on linux as well as on windows including the full deployment chain (compiling a sketch in theia and deploying it over the deployment-server/device-connector)

**Comment by JonasHelming** [comment]
Using docker-compose 1.25.0
jonas@onefour:~/tmp/cdtclouddeploymenttest/cdtcloud-deploymentserver$ docker-compose up db
ERROR: Version in "./docker-compose.yml" is unsupported. You might be seeing this error because you're using the wrong Compose file version. Either specify a supported version (e.g "2.2" or "3.3") and place your service definitions under the `services` key, or omit the `version` key and place your service definitions at the root of the file to use version 1.
For more on the Compose file format versions, see https://docs.docker.com/compose/compose-file/

**Comment by Gozzim** [comment]
> Using docker-compose 1.25.0 jonas@onefour:~/tmp/cdtclouddeploymenttest/cdtcloud-deploymentserver$ docker-compose up db ERROR: Version in "./docker-compose.yml" is unsupported. You might be seeing this error because you're using the wrong Compose file version. Either specify a supported version (e.g "2.2" or "3.3") and place your service definitions under the `services` key, or omit the `version` key and place your service definitions at the root of the file to use version 1. For more on the Compose file format versions, see https://docs.docker.com/compose/compose-file/

The docker-compose version specified within the file was "3.9"
I set it to "3" now, to allow for more room. But Any version below 3 might cause errors as well due to the syntax.
I can of course change it to support version 1 but that is deprecated and will cause issues with higher versions from what I understand.

**Comment by JonasHelming** [comment]
I am pretty sure, that the docker-compose file version does not match one-to-one to the version of docker-compose as a tool. I think you should better do a quick research on this though. My version is what I get from apt, so should be a common one (unless you tell me otherwise).

with the new adaptation, I get:

ERROR: The Compose file './docker-compose.yml' is invalid because:
Unsupported config option for services.arduino-cli: 'profiles'
Unsupported config option for services.db: 'profiles'
Unsupported config option for services.demo: 'profiles'
Unsupported config option for services.deployment-server: 'profiles'
Unsupported config option for services.device-connector: 'profiles'
Unsupported config option for services.theia: 'profiles'
Invalid top-level property "x-networks". Valid top-level sections for this Compose file are: version, services, networks, volumes, and extensions starting with "x-".

You might be seeing this error because you're using the wrong Compose file version. Either specify a supported version (e.g "2.2" or "3.3") and place your service definitions under the `services` key, or omit the `version` key and place your service definitions at the root of the file to use version 1.
For more on the Compose file format versions, see https://docs.docker.com/compose/compose-file/
services.arduino-cli.build contains unsupported option: 'cache_from'
services.demo.build contains unsupported option: 'cache_from'
services.theia.build contains unsupported option: 'cache_from'
services.deployment-server.build contains unsupported option: 'cache_from'
services.device-connector.build contains unsupported option: 'cache_from'

**Comment by JonasHelming** [comment]
error deployment-server@1.0.0: The engine "node" is incompatible with this module. Expected version ">=16.5.0". Got "14.19.1"
error Commands cannot run with an incompatible environment.


(on Import DB)

Do we really require Node 16? Can we make it run on Node 14, e.g. 14.19.1

**Comment by Gozzim** [comment]
> Do we really require Node 16? Can we make it run on Node 14, e.g. 14.19.1

I'm not quite sure of this. I am setting up a completely clean machine now to be able to test everything there.
I'm going to look into this.

**Comment by Gozzim** [comment]
Due to undici in the deployment-server and the device-connector, they require at least version 16. Thanks to @WoH for pointing this out.
We would have to use a different module if we wanted to support version 14.
The install dependencies script will upgrade the node-version to 17 though. So executing this first will also allow to run everything without having to manually install another node version.

**Comment by JonasHelming** [comment]
What is the initial dependency script? Also what I do not get: Shouldn't ./cdtcloud be executed IN a docker container?

**Comment by Gozzim** [comment]
The initial dependency script installs the requirements to execute the services including the ones needed for native execution on the system.
The ./cdtcloud script is just a little summary of commands to improve usability and make it easier for the user. The commands could just be executed manually as well.

**Comment by theZasa** [comment]
Testing on windows on a freshly cloned repo. 

Currently the line  `curl -fsSL https://deno.land/x/install/install.sh | DENO_INSTALL="$PATH_DEPS/deno" sh`
adding the -k parameter seems to resolve the issue:
`curl -fsSL -k https://deno.land/x/install/install.sh | DENO_INSTALL="$PATH_DEPS/deno" sh`

Following the script by running the start:demo command. The script now runs for ca. 5 minutes and then launches all services including theia and the dashboard. 

As mentioned in the PR the arduino-cli in docker on windows does not recognize attached devices due to the privileged option not being supported. I could again reproduce this.

File: apps/shared/deno.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+function denoInstall() {
+
+    { # try
+        echo "Deno version check:" && denoCmd upgrade --version $DENO_MIN_VERSION
+    } ||
+    { # catch
+        echo "Installing Deno..."
+        # just one line of command that works on all OSes
+        # (temporary cd into PATH_DEPS)
+        curl -fsSL https://deno.land/x/install/install.sh | DENO_INSTALL="$PATH_DEPS/deno" sh
```
Comment (theZasa) [inline apps/shared/deno.sh:12]: add -k

Comment (Gozzim) [reply apps/shared/deno.sh]: Added as temporary Hotfix in 686535d7a4ddea809331ba7635d7193015fd2121

**Review by theZasa — APPROVED** [review]

**Pull request merged**